test/spawn.jl

```diff
@@ -173,14 +173,17 @@ exename = joinpath(JULIA_HOME, Base.julia_exename())
 if valgrind_off
     # If --trace-children=yes is passed to valgrind, we will get a
     # valgrind banner here, not "Hello World\n".
-    @test readall(pipe(`$exename -f -e 'println(STDERR,"Hello World")'`, stderr=`cat`)) == "Hello World\n"
+    @test readall(pipeline(`$exename -f -e 'println(STDERR,"Hello World")'`, stderr=`cat`)) == "Hello World\n"
+    p = Pipe()
+    run(pipeline(`$exename -f -e 'println(STDERR,"Hello World")'`, stderr = p))
```

Comment:
the problem i have with this API is that it looks like you could use `p` repeated to aggregate the stderr for many processes, but in fact you can't even call `run` twice on this `Cmd` object, since `run` mutates the `p` object when it opens it. What is the disadvantage to making `Pipe` refer to the entire pipe, while adding an internal type that represents just one end of the pipe (e.g. what is called Pipe in this PR)?